### PR TITLE
Added param-not-constant check to invocation tracker

### PR DESCRIPTION
**Pull request by duncanp-sonar**
Modified _MethodSignatureHelper_ to handle method overrides
Added unit tests for _MethodSignatureHelper_

Test infrastructure:
* added _SnippetCompiler_ class
* refactored _SymbolHelperTests_ to use the new class

### CODE REVIEW

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpInvocationTracker.cs [hunk 1]
```diff
@@ -47,7 +46,19 @@ public CSharpInvocationTracker(IAnalyzerConfiguration analysisConfiguration, Dia
         public override InvocationCondition MatchSimpleNames(params MethodSignature[] methods)
         {
             return (context) => MethodSignatureHelper.IsMatch(context.Identifier as SimpleNameSyntax,
-                    context.Model, context.InvokedMethodSymbol, methods);
+                    context.Model, context.InvokedMethodSymbol, true, methods);
+        }
+
+        public override bool FirstParameterIsStringAndIsNotConstant(InvocationContext context)
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/CSharpInvocationTracker.cs]: New check required by the "executing XPath" rule.
I've added an _IsConstant_ extension method to the _xxxSyntaxHelper_ classes.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/KnownType.cs [hunk 2]
```diff
@@ -392,6 +392,12 @@ internal sealed class KnownType
 
         private readonly bool isSpecialType;
         private readonly SpecialType specialType;
+
+        static internal KnownType CreateKnownTypeForTesting(string typeName)
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/src/SonarAnalyzer.Common/Helpers/KnownType.cs]: I was going to delete this because in the end I didn't need it for the tests I've created.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/SymbolHelperTest.cs [hunk 3]
```diff
@@ -73,193 +73,120 @@ public interface IInterface
...
-
-            this.baseClassDeclaration = this.tree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>()
-                .First(m => m.Identifier.ValueText == "Base");
-            this.derivedClassDeclaration1 = this.tree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>()
-                .First(m => m.Identifier.ValueText == "Derived1");
-            this.derivedClassDeclaration2 = this.tree.GetRoot().DescendantNodes().OfType<ClassDeclarationSyntax>()
-                .First(m => m.Identifier.ValueText == "Derived2");
-            this.interfaceDeclaration = this.tree.GetRoot().DescendantNodes().OfType<InterfaceDeclarationSyntax>()
-                .First(m => m.Identifier.ValueText == "IInterface");
+            this.testCode = new SnippetCompiler(TestInput, ignoreErrors: true, language: AnalyzerLanguage.CSharp);
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/Helpers/SymbolHelperTest.cs]: Refactored an existing test class to use _SnippetCompiler_. No other changes to the tests, or the product code they test.

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs [hunk 4]
```diff
@@ -49,10 +49,18 @@ private ProjectBuilder(Project project)
...
+            var project = parseOptions != null
+                ? Project.WithParseOptions(parseOptions)
+                : Project;
+
+            var compilation = project.GetCompilationAsync().Result;
+
+            return compilationOptions == null
+                ? compilation
+                : compilation.WithOptions(compilationOptions);
+        }
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/ProjectBuilder.cs]: Added the ability to pass in _CompilationOptions_ so the caller can specify e.g. that the assembly is a DLL (to prevent the compile warning about _main method not found_).

File: sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/SnippetCompiler.cs [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+using Microsoft.CodeAnalysis.CSharp;
+using CSharpSyntax = Microsoft.CodeAnalysis.CSharp.Syntax;
+using VBSyntax = Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using Microsoft.CodeAnalysis.VisualBasic;
+
+namespace SonarAnalyzer.UnitTest.TestFramework
+{
+    /// <summary>
+    /// Helper class compiles snippets of code on the fly
```
Comment (duncanp-sonar) [inline sonaranalyzer-dotnet/tests/SonarAnalyzer.UnitTest/TestFramework/SnippetCompiler.cs]: Works for C# and VB.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs [hunk 6]
```diff
@@ -151,6 +151,17 @@ public static SimpleNameSyntax GetIdentifier(this ExpressionSyntax expression)
...
 
+        public static bool IsConstant(this ExpressionSyntax expression, SemanticModel semanticModel)
+        {
+            if (expression == null)
+            {
+                return false;
+            }
+
+            var strippedExpression = expression.RemoveParentheses();
+            return semanticModel.GetConstantValue(strippedExpression).HasValue;
```
[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.VisualBasic/Helpers/VisualBasicSyntaxHelper.cs]: Probably. I'll try it.

File: sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/MethodSignatureHelper.cs [hunk 7]
```diff
@@ -56,7 +56,27 @@ internal static class MethodSignatureHelper
                 {
                     return true;
                 }
+                if (checkOverriddenMethods && IsOverride(m))
+                {
```
[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/MethodSignatureHelper.cs]: In most cases but not all. It'll behave differently if the derived class has added a method with the same name as a method in the base class:

```

  MyBase:Foo(string)

  MyClass:Foo(string)        ' IsOverride will detect this as an override...
  MyClass:Foo(string, int)  ' ... but not this.
```

I'll add another test for this case.

[the remark this answers is not in the document]
Comment (duncanp-sonar) [reply sonaranalyzer-dotnet/src/SonarAnalyzer.CSharp/Helpers/MethodSignatureHelper.cs]: I switched to using _DerivesOrImplements_ and added a test case to document the false-positive corner case.
